sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonFilters.scala

```diff
@@ -0,0 +1,157 @@
+ *    field according to the schema.
+ *    3.1 `skipRow` finds a group of predicates that refers to this JSON field.
+ *    3.2 Per each predicate from the group, `skipRow` decrements its reference counter.
+ *    3.2.1 If predicate reference counter becomes 0, it means that all predicate attributes have
+ *          been already set in the internal row, and the predicate can be applied to it. `skipRow`
+ *          invokes the predicate for the row.
+ *    3.3 `skipRow` applies predicates until one of them returns `false`. In that case, the method
+ *        returns `true` to the parser.
+ *    3.4 If all predicates with zero reference counter return `true`, the final result of
+ *        the method is `false` which tells the parser to not skip the row.
+ * 4. If the parser gets `true` from `JsonFilters.skipRow`, it must not call the method anymore
+ *    for this internal row, and should go the step 1.
+ *
+ * JsonFilters assumes that:
+ *   - reset() is called before any skipRow() calls for new row.
+ *   - skipRow() can be called for any valid index of the struct fields,
+ *      and in any order.
+ *   - After skipRow() returns `true`, the internal state of `JsonFilter` can be inconsistent,
+ *     so, skipRow() must not be called for the current row anymore without reset().
+ *
+ * @param pushedFilters The pushed down source filters. The filters should refer to
+ *                      the fields of the provided schema.
+ * @param schema The required schema of records from datasource files.
+ */
+class JsonFilters(pushedFilters: Seq[sources.Filter], schema: StructType)
+  extends StructFilters(pushedFilters, schema) {
+
+  /**
+   * Stateful JSON predicate that keeps track of its dependent references in the
+   * current row via `refCount`.
+   *
+   * @param predicate The predicate compiled from pushed down source filters.
+   * @param totalRefs The total amount of all filters references which the predicate
+   *                  compiled from.
+   * @param refCount The current number of predicate references in the row that have
+   *                 been not set yet. When `refCount` reaches zero, the predicate
+   *                 has all dependencies are set, and can be applied to the row.
+   */
+  case class JsonPredicate(predicate: BasePredicate, totalRefs: Int, var refCount: Int) {
+    def reset(): Unit = {
+      refCount = totalRefs
+    }
+  }
+
+  // Predicates compiled from the pushed down filters. The predicates are grouped by their
+  // attributes. The i-th group contains predicates that refer to the i-th field of the given
+  // schema. A predicates can be places to many groups if it has many attributes. For example:
+  //  schema: i INTEGER, s STRING
+  //  filters: IsNotNull("i"), AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc"))
+  //  predicates:
+  //    0: Array(IsNotNull("i"), AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc")))
```

Comment:
`AlwaysTrue` is not a good example. It doesn't refer to any attribute and shouldn't be put in any group.

Reply:
I took `AlwaysTrue` because it allows me to highlight a branch in the code.